### PR TITLE
introduces WaitForPodsToStabilizeOnTheSameRevision

### PR DESCRIPTION
**Pull request by p0lyn0mial**
it makes the already existing method `WaitForAPIServerToStabilizeOnTheSameRevision` generic so that it can be used not only for waiting for the apiserver pods.

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
xref: https://github.com/openshift/kubernetes/pull/603

File: test/library/pod_same_revision.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		t.Fatal(err)
+	}
+}
+
+// arePodsOnTheSameRevision tries to find the current revision that the pods are running at.
+// The number of instances is calculated based on the number of running pods in a namespace.
+// This should be okay because this function is meant to be used by WaitForPodsToStabilizeOnTheSameRevision which will wait at least waitForRevisionSuccessThreshold * waitForRevisionSuccessInterval
+// The number of pods should stabilize in that period of time.
+func arePodsOnTheSameRevision(t *testing.T, podClient corev1client.PodInterface, podLabelSelector string) (bool, error) {
+	revisionLabel := "revision"
```
Comment (mfojtik) [inline test/library/pod_same_revision.go:37]: nit: not need for this

File: test/library/pod_same_revision.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(context.TODO(), metav1.ListOptions{LabelSelector: podLabelSelector})
+	if err != nil {
+		// ignore the errors as we hope it will succeed next time
+		t.Logf("failed to list pods, err = %v (this error will be ignored)", err)
+		return false, nil
+	}
+
+	goodRevisions, failingRevisions, progressing, err := getRevisions(revisionLabel, apiServerPods.Items)
+	if err != nil {
```
Comment (mfojtik) [inline test/library/pod_same_revision.go:48]: if err != nil || progressing {
  return false, err
}

File: test/library/pod_same_revision.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		return false, err
+	}
+	if progressing {
+		return false, nil
+	}
+
+	if len(goodRevisions) != 1 {
+		return false, nil // api servers have not converged onto a single revision
+	}
+	revision, _ := goodRevisions.PopAny()
```
Comment (mfojtik) [inline test/library/pod_same_revision.go:58]: nit:
```
if revision, _ := goodRevisions.PopAny(); failingRevisions.Has(revision) {
  return false, fmt.Errorf("api server revision %s has both running and failed pods", revision)
}
```

File: test/library/pod_same_revision.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	return true, nil
+}
+
+func getRevisions(revisionLabel string, pods []corev1.Pod) (sets.String, sets.String, bool, error) {
+	goodRevisions := sets.NewString()
+	badRevisions := sets.NewString()
+
+	if len(pods) == 0 {
```
Comment (mfojtik) [inline test/library/pod_same_revision.go:70]: move this to the top

Comment (mfojtik) [reply test/library/pod_same_revision.go:70]: (or drop this? :-))

**Comment by mfojtik** [comment]
just nits, LGTM otherwise, @sttts to approve :-)

File: test/library/pod_same_revision.go (hunk 1 above)
Comment (p0lyn0mial) [reply test/library/pod_same_revision.go:37]: used in `getRevisions` function

File: test/library/pod_same_revision.go (hunk 4 above)
Comment (p0lyn0mial) [reply test/library/pod_same_revision.go:60]: done

**Pull request merged**